### PR TITLE
Use of std::invoke to run the body in algorithms and graph nodes

### PR DESCRIPTION
**Pull request by kboyarinov**
### Description 
Change the execution of the body in TBB algorithms and Flow Graph nodes to use ``std::invoke`` instead of ``operator()``.
It allows to pass pointers to non-static member functions and non-static member objects as a body.
Affected components:

- ``parallel_for`` 
- ``parallel_reduce``
- ``parallel_deterministic_reduce``
- ``parallel_for_each``
- ``parallel_scan``
- ``parallel_pipeline``
- ``function_node``
- ``multifunction_node``
- ``async_node``
- ``sequencer_node``
- ``join_node`` with ``key_matching`` policy

Fixes # - _issue number(s) if exists_

- [ ] - git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [ ] bug fix - _change that fixes an issue_
- [x] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [x] added - _required for new features and some bug fixes_
- [ ] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [x] needs to be updated
- [ ] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
_List users with `@` to send notifications_

### CODE REVIEW

File: include/oneapi/tbb/parallel_for.h [hunk 1]
```diff
@@ -310,7 +305,7 @@ void parallel_for_impl(Index first, Index last, Index step, const Function& f, P
         throw_exception(exception_id::nonpositive_step); // throws std::invalid_argument
     else if (first < last) {
         // Above "else" avoids "potential divide by zero" warning on some platforms
-        Index end = (last - first - Index(1)) / step + Index(1);
+        Index end = Index(last - first - 1ul) / step + Index(1);
```
Comment (vossmjp) [inline include/oneapi/tbb/parallel_for.h:308]: This seems unrelated to std::invoke.

**Review by vossmjp — COMMENTED** [review]
Partial review. So far have looked at _utils.h, parallel_for.h, parallel_for_each.h, parallel_scan.h, _pipeline_filters.h and parallel_reduce.h. Mostly looks good to me. A couple of small comments/questions. Still need to look at tests, conformance tests & flow graph.

File: include/oneapi/tbb/parallel_for.h (hunk 1 above)
Comment (kboyarinov) [reply include/oneapi/tbb/parallel_for.h:308]: Yes, it is not related for `std::invoke`, but I failed to implement custom `ParallelForIndex` as it defined in the spec to make this line work. I faced a lot of problems with the operation priority, conversions and wrong return values for mathematical operations. This workaround can be considered as partial fix for the implementation to better match the spec.

File: include/oneapi/tbb/detail/_pipeline_filters.h [hunk 2]
```diff
@@ -281,7 +281,7 @@ class concrete_filter<InputType, void, Body>: public base_filter {
 
     void* operator()(void* input) override {
         input_pointer temp_input = input_helper::cast_from_void_ptr(input);
-        my_body(std::move(input_helper::token(temp_input)));
+        tbb::detail::invoke(my_body, std::move(input_helper::token(temp_input)));
```
[the remark this answers is not in the document]
Comment (kboyarinov) [reply include/oneapi/tbb/detail/_pipeline_filters.h:284]: Yes, it is - to have any profit from using `std::invoke`, the user will need to provide pointer-to-member function or member object from the input object. And currently we don't have any useful functions in `flow_control` that can be used as as a filter body.

File: include/oneapi/tbb/detail/_pipeline_filters_deduction.h [hunk 3]
```diff
@@ -26,18 +26,18 @@ namespace detail {
 namespace d1 {
 
 template <typename Input, typename Output>
-struct declare_fitler_types {
+struct declare_filter_types {
```
Comment (omalyshe) [inline include/oneapi/tbb/detail/_pipeline_filters_deduction.h:29]: Copyright year should be fixed

File: include/oneapi/tbb/parallel_scan.h [hunk 4]
```diff
@@ -1,6 +1,5 @@
 /*
-    Copyright (c) 2005-2021 Intel Corporation
-
+    Copyright (c) 2005-2023 Intel Corporation
```
Comment (omalyshe) [inline include/oneapi/tbb/parallel_scan.h:2]: Insert the empty line after Copyright

File: test/tbb/test_task.cpp [hunk 5]
```diff
@@ -785,7 +785,7 @@ TEST_CASE("Test with priority inversion") {
     // take first core on execute
     utils::SpinBarrier barrier(thread_number + 1);
     test_arena.execute([&] {
-        tbb::parallel_for(std::uint32_t(0), thread_number + 1, [&] (std::uint32_t&) {
+        tbb::parallel_for(std::uint32_t(0), thread_number + 1, [&] (std::uint32_t) {
```
Comment (omalyshe) [inline test/tbb/test_task.cpp:788]: Copyright year should be fixed

File: include/oneapi/tbb/detail/_pipeline_filters_deduction.h (hunk 3 above)
Comment (kboyarinov) [reply include/oneapi/tbb/detail/_pipeline_filters_deduction.h:29]: Done

File: test/tbb/test_task.cpp (hunk 5 above)
Comment (kboyarinov) [reply test/tbb/test_task.cpp:788]: Done

File: include/oneapi/tbb/parallel_scan.h (hunk 4 above)
Comment (kboyarinov) [reply include/oneapi/tbb/parallel_scan.h:2]: Done

File: include/oneapi/tbb/detail/_flow_graph_body_impl.h [hunk 6]
```diff
@@ -218,7 +218,7 @@ template <typename Input, typename Output, typename B>
 class type_to_key_function_body_leaf : public type_to_key_function_body<Input, Output> {
 public:
     type_to_key_function_body_leaf( const B &_body ) : body(_body) { }
-    Output operator()(const Input &i) override { return body(i); }
+    Output operator()(const Input &i) override { return tbb::detail::invoke(body, i); }
```
Comment (vossmjp) [inline include/oneapi/tbb/detail/_flow_graph_body_impl.h:221]: I suppose you intend to exclude use of the invoke from the Output& specialization's operator(), since it would used to return a reference to something in the input message, which would be unsafe. Is that right?

File: include/oneapi/tbb/detail/_flow_graph_node_impl.h [hunk 7]
```diff
@@ -496,7 +496,7 @@ class multifunction_input : public function_input_base<Input, Policy, A, multifu
     // constructor
     template<typename Body>
     multifunction_input(graph &g, size_t max_concurrency,Body& body, node_priority_t a_priority )
-      : base_type(g, max_concurrency, a_priority, noexcept(body(input_type(), my_output_ports)))
+      : base_type(g, max_concurrency, a_priority, noexcept(tbb::detail::invoke(body, input_type(), my_output_ports)))
```
Comment (vossmjp) [inline include/oneapi/tbb/detail/_flow_graph_node_impl.h:499]: I suppose you intend to exclude the use of invoke in the apply_body_impl_bypass. Is there a chance that the input message could define a function that would accept the output ports, and so be a valid use case?

**Review by vossmjp — COMMENTED** [review]
The flow graph modifications look ok to me. Made a couple of comments to make sure I understand why certain invocation points are excluded.

File: include/oneapi/tbb/detail/_flow_graph_body_impl.h (hunk 6 above)
Comment (kboyarinov) [reply include/oneapi/tbb/detail/_flow_graph_body_impl.h:221]: Actually, I missed that use-case. I checked for its safety and I don't see any unsafe actions, because our interface forces the body to return constant reference for some object inside the input message and never modifies it. I have updated the implementation and added the test for it.

File: include/oneapi/tbb/detail/_flow_graph_node_impl.h (hunk 7 above)
Comment (kboyarinov) [reply include/oneapi/tbb/detail/_flow_graph_node_impl.h:499]: I have excluded `apply_body_impl_bypass` because we call operator() of `multifunction_body_leaf` there and it is always a callable object defined inside oneTBB. It supports `std::invoke` inside, so use-case you described would work and we have a test for it.

**Review by vossmjp — APPROVED** [review]
Did a quick look through tests and they look ok to me.

**Review by pavelkumbrasev — APPROVED** [review]
